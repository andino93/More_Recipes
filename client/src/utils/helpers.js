import jwtDecode from 'jwt-decode';
import history from '../utils/history';

const decodeToken = token => jwtDecode(token);

const logout = () => {
  window.localStorage.removeItem('token');
  history.push('/');
};

const authorization = () => ({
  headers: {
    'x-access-token': window.localStorage.token
  }
});

const recipeResponseType = {
  FAILURE: 0,
  ADD_RECIPE_SUCCESS: 1,
  EDIT_RECIPE_SUCCESS: 2,
  VIEW_RECIPE_SUCCESS: 3,
  GET_REVIEW_SUCCESS: 4,
  POST_REVIEW_SUCCESS: 5,
  CLEAR_RECIPE_MESSAGE: 6,
  DELETE_RECIPE_SUCCESS: 7,
  GET_ALL_RECIPES_SUCCESS: 9,
  CREATE_USER_FAVOURITE: 10,
  GET_USER_FAVORITE_SUCCESS: 11,
  CREATE_UPVOTE_SUCCESSFUL: 12,
  CREATE_DOWNVOTE_SUCCESSFUL: 13,
  GET_ALL_FAVORITES_SUCCESS: 14,
  SEARCH_RECIPE_SUCCESS: 15,
  GET_USER_RECIPES_SUCCESS: 16
};

const userResponseType = {
  FAILURE: 0,
  GET_USER_PROFILE_SUCCESS: 1
};

export {
  recipeResponseType,
  userResponseType,
  decodeToken,
  logout,
  authorization
};
