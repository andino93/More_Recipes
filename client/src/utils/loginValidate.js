import Validator from 'validator';
import isEmpty from 'lodash/isEmpty';

const loginValidate = (data) => {
  const errors = {};

  if (Validator.isEmpty(data.email)) {
    errors.email = 'Email is required';
  }

  if (Validator.isEmpty(data.password)) {
    errors.password = 'Password is required';
  }

  return {
    errors,
    isValid: isEmpty(errors)
  };
};

export default loginValidate;
