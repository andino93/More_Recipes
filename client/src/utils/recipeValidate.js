import Validator from 'validator';
import isEmpty from 'lodash/isEmpty';

const recipeValidate = (data) => {
  const name = data.name.trim(),
    description = data.description.trim(),
    ingredient = data.ingredient.trim();
  const errors = {};

  if (!Validator.isEmpty(name)) {
    const hasNumber = name.split('').filter(character => Validator.toInt(character));
    if (hasNumber.length !== 0) {
      errors.name = 'Recipe name must contain alphabetic characters only';
    }
  } else { errors.name = 'Recipe name is required'; }

  if (!Validator.isEmpty(description)) {
    if (!Validator.isLength(description, { min: 4, max: undefined })) {
      errors.description = 'Recipe description must be at least 4 characters';
    }
  } else { errors.description = 'Recipe description is required'; }

  if (!Validator.isEmpty(ingredient)) {
    if (!Validator.isLength(ingredient, { min: 3, max: undefined })) {
      errors.ingredient = 'Recipe ingredient must be at least 3 characters';
    }
  } else { errors.ingredient = 'Recipe ingredient is required'; }

  return {
    errors,
    isValid: isEmpty(errors)
  };
};

export default recipeValidate;