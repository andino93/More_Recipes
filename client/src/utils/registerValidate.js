import Validator from 'validator';
import isEmpty from 'lodash/isEmpty';

const registerValidate = (data) => {
  const errors = {};
  
  if (Validator.isEmpty(data.firstName)) {
    errors.firstName = 'First name is required';
  }

  if (Validator.isEmpty(data.lastName)) {
    errors.lastName = 'Last name is required';
  }

  if (Validator.isEmpty(data.username)) {
    errors.username = 'Username is required';
  }

  if (Validator.isEmpty(data.email)) {
    errors.email = 'Email is required';
  }

  if (Validator.isEmpty(data.password)) {
    errors.password = 'Password is required';
  }

  if (Validator.isEmpty(data.confirmPassword)) {
    errors.confirmPassword = 'Confirm password is required';
  }

  if (!Validator.equals(data.password, data.confirmPassword)) {
    errors.confirmPassword = 'Passwords must match';
  }

  return {
    errors,
    isValid: isEmpty(errors)
  };
};

export default registerValidate;
