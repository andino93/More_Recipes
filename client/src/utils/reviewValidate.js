import Validator from 'validator';
import isEmpty from 'lodash/isEmpty';

const reviewValidate = (data) => {
  const content = data.content.trim();
  const errors = {};

  if (!Validator.isEmpty(content)) {
    if (!Validator.isLength(content, { min: 4, max: undefined })) {
      errors.content = 'Review  must be at least 4 characters';
    }
  } else { errors.content = 'Review details is required'; }

  return {
    errors,
    isValid: isEmpty(errors)
  };
};

export default reviewValidate;