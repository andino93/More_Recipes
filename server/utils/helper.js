const errorHandler = (error, res) => {
  switch (error.code) {
    case 404:
      return res.status(404).json({ message: error.message });
    default:
      return res.status(500).json(error);
  }
};


const generatePaginationMeta = (dbResult, limit, page) => {
  const paginationMeta = {
    pageCount: Math.ceil(dbResult.count / limit),
    totalCount: dbResult.count,
    outputCount: dbResult.rows.length,
    pageSize: limit,
    currentPage: page + 1
  };
  return paginationMeta;
};

export { errorHandler, generatePaginationMeta };
