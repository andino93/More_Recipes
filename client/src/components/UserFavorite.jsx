import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'react-proptypes';
import ReactPaginate from 'react-paginate';
import { getUserFavorites } from '../actions/recipe/recipeActions';
import { decodeToken, recipeResponseType } from '../utils/helpers';
import Header from './partials/Headers/Header.jsx';
import emptyLogo from '../../src/build/assets/images/Empty.png';
import FooterComponent from './partials/Footer.jsx';
import FavoriteRecipe from './recipe/FavoriteRecipe.jsx';

/**
 *
 * @class UserFavorite
 * 
 * @extends {React.Component}
 */
class UserFavorite extends Component {
  /**
   *
   * @param {props} props
   */
  constructor(props) {
    super(props);
    this.state = {
      favorites: [],
      pageCount: 0,
      page: 1
    };
    this.handlePageChange = this.handlePageChange.bind(this);
  }
  /**
   * @returns {XML} XML/JSX
   * 
   * @memberof UserFavorite
   */
  componentWillMount() {
    const { id } = decodeToken(window.localStorage.token);
    const { page } = this.state;
    this.props.getUserFavorites(id, page);
  }
  /**
   * @param  {object} nextProps
   * 
   * @returns {XML} XML/JSX
   * 
   * @memberof UserFavorites
   */
  componentWillReceiveProps(nextProps) {
    const { responseType, favorites, pagination } = nextProps.favoriteState;
    if (responseType === recipeResponseType.GET_USER_FAVORITE_SUCCESS) {
      this.setState({
        favorites,
        pageCount: pagination.pageCount || 0
      });
    }
  }
  /**
   * Handles page change
   *
   * @method handlePageChange
   *
   * @param {event} page
   *
   * @return {void}
   */
  handlePageChange(page) {
    const currentPage = page.selected + 1;
    const { id } = decodeToken(window.localStorage.token);
    this.props.getUserFavorites(id, currentPage);
  }
  /**
     * 
     * @returns {XML} XML/JSX
     * 
     * @memberof UserFavorite
     */
  render() {
    const { favorites } = this.state;
    return (
      <div>
        <Header/>
        <div className="wrapper">
          <div className="container">
            <nav aria-label="breadcrumb" className="mt-40 mb-10" >
              <ol className="breadcrumb bg-white">
                <li className="breadcrumb-item"><Link to="/">Home</Link></li>
                <li className="breadcrumb-item"><Link to="/profile">
                User</Link></li>
                <li className="breadcrumb-item active" aria-current="page">
                My Favorites</li>
              </ol>
            </nav>
          </div>
          <div className="container mb-20 recipe-details-container">
            <div className="row">
              {
                favorites && favorites.length > 0 ?
                  favorites.map((favorite, index) => (
                    <FavoriteRecipe
                      key={index}
                      recipe={favorite.Recipe}
                    />
                  ))
                  :
                  <div className="col col-md-12 text-center">
                    <h1 className="text-info">You have no favorites recipes yet!</h1>
                    <img src={emptyLogo} alt="No Recipe"/>
                  </div>
              }
            </div>
          </div>
          <div className="clearfix" />
          {
            favorites && favorites.length > 0 ?
              <ReactPaginate
                pageCount={this.state.pageCount}
                pageRangeDisplayed={2}
                marginPagesDisplayed={2}
                onPageChange={this.handlePageChange}
                containerClassName="pagination justify-content-center"
                subContainerClassName="page-item"
                pageLinkClassName="page-link"
                activeClassName="page-item active"
                nextLinkClassName="page-link"
                previousLinkClassName="page-link"
              /> : null
          }
        </div>
        <FooterComponent/>
      </div>
    );
  }
}

UserFavorite.propTypes = {
  getUserFavorites: PropTypes.func.isRequired,
  favoriteState: PropTypes.object
};

const mapStateToProps = state => ({
  favoriteState: state.recipeReducer
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ getUserFavorites }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(UserFavorite);