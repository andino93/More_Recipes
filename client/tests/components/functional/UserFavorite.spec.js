import React from 'react';
import expect from 'expect';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { configure } from 'enzyme';
import { recipeResponseType } from '../../../src/utils/helpers';
import mockStore from '../../__mocks__/mockStore';
import mockData from '../../__mocks__/mockData';
import mountWrap from '../../__mocks__/utils';
import localStorageMock from '../../__mocks__/localStorageMock';
import UserFavorite from '../../../src/components/UserFavorite.jsx';

configure({ adapter: new Adapter() });

// noinspection JSAnnotator
window.localStorage = localStorageMock;
window.localStorage.token = mockData.token;

describe('<UserFavorite>', () => {
  const store = mockStore({});
  const connectedUserFavorite = mountWrap(UserFavorite, store, {
    match: {}
  });
  it('should render the Profile Component', (done) => {
    expect(toJson(connectedUserFavorite)).toMatchSnapshot();
    done();
  });
  it('should check for componentWillReceiveProps', () => {
    const { userFavorites: { favorites } } = mockData;
    connectedUserFavorite.find('UserFavorite').instance().componentWillReceiveProps({
      favoriteState: {
        responseType: recipeResponseType.GET_USER_FAVORITE_SUCCESS,
        pagination: {
          pageCount: 10
        },
        favorites
      },
    });
  });
  it('should simulate setState', (done) => {
    connectedUserFavorite.find('UserFavorite').instance().setState({
      favorites: [],
      pageCount: 1
    });
    done();
  });
});