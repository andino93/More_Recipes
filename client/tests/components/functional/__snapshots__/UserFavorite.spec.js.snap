// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserFavorite> should render the Profile Component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "blank",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(UserFavorite)
        match={Object {}}
      >
        <UserFavorite
          getUserFavorites={[Function]}
          match={Object {}}
        >
          <div>
            <Connect(Header)>
              <Header
                logoutAction={[Function]}
              >
                <div>
                  <nav
                    className="navbar navbar-expand-sm navbar-light bg-navbar fixed-top"
                  >
                    <Link
                      className="navbar-brand"
                      replace={false}
                      to="/"
                    >
                      <a
                        className="navbar-brand"
                        href="/"
                        onClick={[Function]}
                      >
                        More Recipes
                      </a>
                    </Link>
                    <button
                      aria-controls="navbarSupportedContent"
                      aria-expanded="false"
                      aria-label="Toggle navigation"
                      className="navbar-toggler"
                      data-target="#navbarSupportedContent"
                      data-toggle="collapse"
                      type="button"
                    >
                      <span
                        className="navbar-toggler-icon"
                      />
                    </button>
                    <div
                      className="collapse navbar-collapse"
                      id="navbarSupportedContent"
                    >
                      <ul
                        className="navbar-nav ml-auto"
                      >
                        <li
                          className="nav-item dropdown"
                        >
                          <Link
                            aria-expanded="false"
                            aria-haspopup="true"
                            className="nav-link dropdown-toggle"
                            data-toggle="dropdown"
                            id="navbarDropdown"
                            replace={false}
                            role="button"
                            to="#"
                          >
                            <a
                              aria-expanded="false"
                              aria-haspopup="true"
                              className="nav-link dropdown-toggle"
                              data-toggle="dropdown"
                              href="#"
                              id="navbarDropdown"
                              onClick={[Function]}
                              role="button"
                            >
                              My Account
                            </a>
                          </Link>
                          <div
                            aria-labelledby="navbarDropdown"
                            className="dropdown-menu dropdown-menu-right mr-20"
                          >
                            <Link
                              className="dropdown-item"
                              replace={false}
                              to="/profile"
                            >
                              <a
                                className="dropdown-item"
                                href="/profile"
                                onClick={[Function]}
                              >
                                <i
                                  className="fa fa-user-circle"
                                />
                                 My Profile
                              </a>
                            </Link>
                            <Link
                              className="dropdown-item"
                              replace={false}
                              to="/user-recipes"
                            >
                              <a
                                className="dropdown-item"
                                href="/user-recipes"
                                onClick={[Function]}
                              >
                                <i
                                  className="fa fa-list"
                                />
                                 My Recipes
                              </a>
                            </Link>
                            <Link
                              className="dropdown-item"
                              replace={false}
                              to="/my-favorite"
                            >
                              <a
                                className="dropdown-item"
                                href="/my-favorite"
                                onClick={[Function]}
                              >
                                <i
                                  className="fa fa-star"
                                />
                                 My Favorites
                              </a>
                            </Link>
                            <Link
                              className="dropdown-item"
                              replace={false}
                              to="/addRecipe"
                            >
                              <a
                                className="dropdown-item"
                                href="/addRecipe"
                                onClick={[Function]}
                              >
                                <i
                                  className="fa fa-plus"
                                />
                                 Add Recipes
                              </a>
                            </Link>
                            <div
                              className="dropdown-divider"
                            />
                            <Link
                              className="dropdown-item"
                              onClick={[Function]}
                              replace={false}
                              to="#"
                            >
                              <a
                                className="dropdown-item"
                                href="#"
                                onClick={[Function]}
                              >
                                <i
                                  className="fa fa-sign-out"
                                />
                                Log out
                              </a>
                            </Link>
                          </div>
                        </li>
                      </ul>
                    </div>
                  </nav>
                </div>
              </Header>
            </Connect(Header)>
            <div
              className="wrapper"
            >
              <div
                className="container"
              >
                <nav
                  aria-label="breadcrumb"
                  className="mt-40 mb-10"
                >
                  <ol
                    className="breadcrumb bg-white"
                  >
                    <li
                      className="breadcrumb-item"
                    >
                      <Link
                        replace={false}
                        to="/"
                      >
                        <a
                          href="/"
                          onClick={[Function]}
                        >
                          Home
                        </a>
                      </Link>
                    </li>
                    <li
                      className="breadcrumb-item"
                    >
                      <Link
                        replace={false}
                        to="/profile"
                      >
                        <a
                          href="/profile"
                          onClick={[Function]}
                        >
                          User
                        </a>
                      </Link>
                    </li>
                    <li
                      aria-current="page"
                      className="breadcrumb-item active"
                    >
                      My Favorites
                    </li>
                  </ol>
                </nav>
              </div>
              <div
                className="container mb-20 recipe-details-container"
              >
                <div
                  className="row"
                >
                  <div
                    className="col col-md-12 text-center"
                  >
                    <h1
                      className="text-info"
                    >
                      You have no favorites recipes yet!
                    </h1>
                    <img
                      alt="No Recipe"
                      src="test-file-stub"
                    />
                  </div>
                </div>
              </div>
              <div
                className="clearfix"
              />
            </div>
            <FooterComponent>
              <footer
                className="p-10 well well-sm m-0 shadow-lite"
              >
                <div
                  className="text-center text-muted"
                >
                  <p
                    className="bold"
                  >
                    © 2017 More Recipes
                  </p>
                </div>
              </footer>
            </FooterComponent>
          </div>
        </UserFavorite>
      </Connect(UserFavorite)>
    </Router>
  </BrowserRouter>
</Provider>
`;
